cbits/fpstring.c

```diff
@@ -113,3 +107,160 @@ int fps_compare(const void *a, const void *b) {
 void fps_sort(unsigned char *p, size_t len) {
   return qsort(p, len, 1, fps_compare);
 }
+
+/* count the number of occurences of a char in a string */
+size_t fps_count_naive(unsigned char *str, size_t len, unsigned char w) {
+    size_t c;
+    for (c = 0; len-- != 0; ++str)
+        if (*str == w)
+            ++c;
+    return c;
+}
+
+#if defined(__x86_64__) && (__GNUC__ >= 6 || defined(__clang_major__))
+#define USE_SIMD_COUNT
+#endif
+
+#ifdef USE_SIMD_COUNT
+__attribute__((target("sse4.2")))
+size_t fps_count_cmpestrm(unsigned char *str, size_t len, unsigned char w) {
+    const __m128i pat = _mm_set1_epi8(w);
+
+    size_t res = 0;
+
+    size_t i = 0;
+
+    for (; i < len && (intptr_t)(str + i) % 64; ++i) {
+        res += str[i] == w;
+    }
+
+    for (size_t end = len - 128; i < end; i += 128) {
+        __m128i p0 = _mm_load_si128((const __m128i*)(str + i + 16 * 0));
+        __m128i p1 = _mm_load_si128((const __m128i*)(str + i + 16 * 1));
+        __m128i p2 = _mm_load_si128((const __m128i*)(str + i + 16 * 2));
+        __m128i p3 = _mm_load_si128((const __m128i*)(str + i + 16 * 3));
+        __m128i p4 = _mm_load_si128((const __m128i*)(str + i + 16 * 4));
+        __m128i p5 = _mm_load_si128((const __m128i*)(str + i + 16 * 5));
+        __m128i p6 = _mm_load_si128((const __m128i*)(str + i + 16 * 6));
+        __m128i p7 = _mm_load_si128((const __m128i*)(str + i + 16 * 7));
+#define MODE _SIDD_SBYTE_OPS | _SIDD_CMP_EQUAL_EACH
```

Comment:
Please add the reference documentation link https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=1885,6090,5596,5653,4115,4076,4115,4938,4956,4115,5608,5656,5608,835,2426,835&text=_mm_cmpestrm for the next person who doesn't know what `_mm_cmpestrm` is.